libraries/AP_HAL_ChibiOS/UARTDriver.cpp

```diff
@@ -750,14 +775,39 @@ size_t UARTDriver::write(uint8_t c)
         hal.scheduler->delay(1);
         _write_mutex.take_blocking();
     }
-    size_t ret = _writebuf.write(&c, 1);
+    uint32_t ret = _writebuf.write(&c, 1);
     if (unbuffered_writes) {
         chEvtSignal(uart_thread_ctx, EVT_TRANSMIT_DATA_READY);
     }
     _write_mutex.give();
+
+#if HAL_LOG_UART_STATS
+    {
+        WITH_SEMAPHORE(_stats.sem);
+        _stats.tx_bytes++;
+        update_writebuf_stats();
+    }
+#endif
     return ret;
 }
 
+void UARTDriver::update_writebuf_stats()
+{
+    const uint32_t space = _writebuf.space();
+    _stats.tx_buffer_min = MIN(_stats.tx_buffer_min, space);
+    _stats.tx_buffer_max = MAX(_stats.tx_buffer_max, space);
+    _stats.tx_buffer_bytes_space_sum += space;
+    _stats.tx_buffer_bytes_space_count++;
+}
+void UARTDriver::update_readbuf_stats()
+{
+    const uint32_t available = _readbuf.available();
```

Comment:
It feels like it's easier to add the semaphore holding in here, rather then making all the callers be responsible for having it.